x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/user/TransportGetUserPrivilegesAction.java

```diff
@@ -0,0 +1,132 @@
+import org.elasticsearch.xpack.core.security.authz.permission.FieldPermissionsDefinition;
+import org.elasticsearch.xpack.core.security.authz.permission.IndicesPermission;
+import org.elasticsearch.xpack.core.security.authz.permission.Role;
+import org.elasticsearch.xpack.core.security.authz.privilege.ApplicationPrivilege;
+import org.elasticsearch.xpack.core.security.authz.privilege.ClusterPrivilege;
+import org.elasticsearch.xpack.core.security.authz.privilege.ConditionalClusterPrivilege;
+import org.elasticsearch.xpack.core.security.authz.privilege.Privilege;
+import org.elasticsearch.xpack.core.security.user.User;
+import org.elasticsearch.xpack.security.authz.AuthorizationService;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.LinkedHashSet;
+import java.util.Set;
+import java.util.TreeSet;
+
+import static org.elasticsearch.common.Strings.arrayToCommaDelimitedString;
+
+/**
+ * Transport action for {@link GetUserPrivilegesAction}
+ */
+public class TransportGetUserPrivilegesAction extends HandledTransportAction<GetUserPrivilegesRequest, GetUserPrivilegesResponse> {
+
+    private final ThreadPool threadPool;
+    private final AuthorizationService authorizationService;
+
+    @Inject
+    public TransportGetUserPrivilegesAction(Settings settings, ThreadPool threadPool, TransportService transportService,
+                                            ActionFilters actionFilters, AuthorizationService authorizationService) {
+        super(settings, GetUserPrivilegesAction.NAME, transportService, actionFilters, GetUserPrivilegesRequest::new);
+        this.threadPool = threadPool;
+        this.authorizationService = authorizationService;
+    }
+
+    @Override
+    protected void doExecute(Task task, GetUserPrivilegesRequest request, ActionListener<GetUserPrivilegesResponse> listener) {
+        final String username = request.username();
+
+        final User user = Authentication.getAuthentication(threadPool.getThreadContext()).getUser();
+        if (user.principal().equals(username) == false) {
+            listener.onFailure(new IllegalArgumentException("users may only list the privileges of their own account"));
+            return;
+        }
+
+        authorizationService.roles(user, ActionListener.wrap(
+            role -> listener.onResponse(buildResponseObject(role)),
+            listener::onFailure));
+    }
+
+    // package protected for testing
+    GetUserPrivilegesResponse buildResponseObject(Role userRole) {
+        logger.trace(() -> new ParameterizedMessage("List privileges for role [{}]", arrayToCommaDelimitedString(userRole.names())));
+
+        // We use sorted sets because they will typically be very small, and having a predictable order allows for simpler testing
+        final Set<String> cluster = new TreeSet<>();
+        final Set<ConditionalClusterPrivilege> conditionalCluster = new HashSet<>();
```

Comment:
should we use a sorted set here?

Reply:
For consistency, probably.
The tests don't require it, because the rest tests use `contains` and `length` to check those JSON arrays, but that is only possible for arrays of objects, not strings.

Reply:
The reason I haven't done it, is that it would require defining an ordering for `ConditionalClusterPrivilege`, and it's not obvious what that order would be.

Reply:
can you add this as a code comment for why these are not sorted.